create util function for batched arange

yuguo68 opened the pull request:
Signed-off-by: Yu Guo [yuguo@meta.com](mailto:yuguo@meta.com)

Summary: similar code repeats 3 times, so create a util function.

Differential Revision: D75650703

yuguo68 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -525,17 +543,10 @@ def _prepare_inputs(
-        cumsums_offsets = np.repeat(cu_num_tokens - num_scheduled_tokens,
-                                    num_scheduled_tokens)
-        # Step 3. [0, 1, 0, 1, 2, 3, 4, 0, 1, 2]
-        arange = self.arange_np[:total_num_scheduled_tokens] - cumsums_offsets
```
Yes, total_num_scheduled_tokens is computed here https://github.com/vllm-project/vllm/blob/4d0a1541bed0a89f71c55b0714e18422688bec53/vllm/v1/core/sched/scheduler.py#L483

also for 1D subtraction, the size needs to be the same and np.cumsum(num_scheduled_tokens)[-1] is the correct size for cumsums_offsets.

DarkLight1337 commented:
Thanks!

DarkLight1337 reviewed (approved):
Got it, thanks for cleaning this up!

yeqcharlotte commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -500,6 +500,25 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
+        num_tokens: np.ndarray,
+        cumsum_dtype: Optional[np.dtype] = None,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        # Get batched arange.
+        # E.g., [2, 5, 3] -> [0, 1, 0, 1, 2, 3, 4, 0, 1, 2]
+        # Equivalent to but faster than:
+        # np.concatenate([np.arange(n) for n in num_scheduled_tokens])
```
doc string for function comments?

yuguo68 committed:
create util function for batched arange

Summary:
Signed-off-by: Yu Guo <yuguo@meta.com>

similar code repeats 3 times, so create a util function.

Differential Revision: D75650703

yuguo68 commented:
@DarkLight1337 could you help merge? thanks!
